File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        numsWithoutZero = []
        isZero=False
        countZeros=0
        res=[]
        # check zeros in nums
        for n in nums:
            if n==0:
                countZeros+=1
                if countZeros==2: break
                isZero=True
            else:
                numsWithoutZero.append(n)
        if countZeros==2:
            for n in nums:
                res.append(0)
            return res
        else:
            mlt = 0 if len(numsWithoutZero) == 0 else 1
            for n in numsWithoutZero:
                mlt*=n
        
            for n in nums:
                if isZero:
                    if n!=0:
                        res.append(0)
                    else:
                        res.append(mlt)
                else:
                    res.append(int(mlt/n))
            return res

File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> p(n+1);
        p[0]=1;
        vector<int> s(n+1);
        s[n] = 1;

        for(int i=0;i<n;i++){
            p[i+1] =  p[i] * nums[i];
        }

        for(int i=n-1;i>=0;i--){
           s[i] = s[i+1] * nums[i];
        }

        for(int i=0;i<n;i++){
          nums[i] = p[i] * s[i+1];
        }

        for(int i=0;i<s.size();i++){
           cout<<s[i]<<",";
        }

        return nums;
    }
};